contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -2384,13 +2387,90 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
 
 							if (strcmp(defel->defname, "default_database") == 0)
 								login_options = lappend(login_options, defel);
+							else if (strcmp(defel->defname, "name_location") == 0)
+							{
+								int location = defel->location;
+								orig_loginname = extract_identifier(queryString + location);
+								login_options = lappend(login_options, defel);
+							}
+							else if (strcmp(defel->defname, "from_windows") == 0)
+							{
+								from_windows = true;
+								login_options = lappend(login_options, defel);
+							}
 						}
 
 						foreach(option, login_options)
 						{
 							stmt->options = list_delete_ptr(stmt->options,
 															lfirst(option));
 						}
+
+						if (orig_loginname)
+						{
+							login_options = lappend(login_options, 
+													makeDefElem("original_login_name",
+																(Node *) makeString(orig_loginname),
+																-1));
+						}
+
+						if (from_windows && orig_loginname)
+						{
+							/*
+							* The login name must contain '\' if it is windows login or else throw error.
+							*/
+							if ((strchr(orig_loginname, '\\')) == NULL)
+								ereport(ERROR, (errcode(ERRCODE_INVALID_NAME),
+				 					errmsg("'%s' is not a valid Windows NT name. Give the complete name: <domain\\username>.",
+										orig_loginname)));
+
+							/*
+							*	Check whether login_name has valid length or not.
+							*/
+							if (!check_windows_login_length(orig_loginname))
+								ereport(ERROR,
+									(errcode(ERRCODE_INVALID_NAME),
+									 errmsg("The login name '%s' has invalid length. Login name length should be between %d and %d for windows login.",
+											orig_loginname, (NAMEDATALEN_WINDOWS_MIN + 1), (NAMEDATALEN_WINDOWS_MAX - 1))));
+
+							/*
+							*	Check whether domain name is empty. If the first character is '\', that ensures domain is empty.
```

Comment:
a general comment, I feel we may need to put most of the work in rolecmd.c instead of here. Here is already a big mess.

Reply:
once we make all the changes, we can move all this handling of login to rolecmds under new helper function.